decompiler/pipeline/ssa/dependency_graph.py

```diff
@@ -1,74 +1,64 @@
-from typing import Iterable, List, Optional, Set
+from functools import reduce
+from typing import Iterator
 
 from decompiler.structures.graphs.cfg import ControlFlowGraph
-from decompiler.structures.interferencegraph import InterferenceGraph
+from decompiler.structures.pseudo import Expression, Operation, OperationType
 from decompiler.structures.pseudo.expressions import Variable
 from decompiler.structures.pseudo.instructions import Assignment
-from decompiler.structures.pseudo.operations import Call
-from networkx import DiGraph, weakly_connected_components
+from networkx import MultiDiGraph
 
 
-def _non_call_assignments(cfg: ControlFlowGraph) -> Iterable[Assignment]:
-    """Yield all interesting assignments for the dependency graph."""
-    for instr in cfg.instructions:
-        if isinstance(instr, Assignment) and isinstance(instr.destination, Variable) and not isinstance(instr.value, Call):
-            yield instr
+def dependency_graph_from_cfg(cfg: ControlFlowGraph) -> MultiDiGraph:
+    """
+    Construct the dependency graph of the given CFG, i.e. adds an edge between two variables if they depend on each other.
+        - Add an edge the definition to at most one requirement for each instruction.
+        - All variables that where not defined via Phi-functions before have out-degree of at most 1, because they are defined at most once.
+        - Variables that are defined via Phi-functions can have one successor for each required variable of the Phi-function.
+    """
+    dependency_graph = MultiDiGraph()
+
+    for variable in _collect_variables(cfg):
+        dependency_graph.add_node((variable,))
```

Comment:
I do this, because nodes in the dependency graph are made of tuples of variables. This way, I can more easily merge multiple variables into one node. Because each node starts from one variable, the initialization logic looks like this.